mikeio/res1d.py

```diff
@@ -1,141 +1,274 @@
+                        found_points["reach"].append(reach)
+                    else:
+                        break  # Break at the first chainage found.
 
-        file = self.__read(file_path)
-        dataItemTypes, indices, reachNums = self.find_items(file, queries)
-        df = self._get_data(file, queries, dataItemTypes, indices, reachNums)
-        file.Dispose()
+            if not found_reach:
+                raise DataNotFoundInFile(
+                    f"Reach '{q_reach_name}' was not found.")
+            if not found_chainage:
+                raise DataNotFoundInFile(
+                    f"Chainage {q_chain} was not found.")
+
+        return dict(found_points)
+
+    def read(self, queries):
+        with self.open() as self.file:
+            found_points = self._find_points(queries)
+            df = self._get_data(found_points)
         return df
 
 
-class ExtractionPoint:
-    def VariableType(VariableType):
-        """
-            Variable Type (eg. WaterLevel or Discharge)
-        """
-        return VariableType
+class QueryData:
+    """A query object that declares what data should be
+    extracted from a .res1d file.
+    
+    Parameters
+    ----------
+    variable_type: str
+        Either 'WaterLevel', 'Discharge' or 'Pollutant'
+    branch_name: str, optional
+        Branch name, consider all the branches if None
+    chainage: float, optional
+        chainage, considers all the chainages if None
+    
+    Examples
+    --------
+    `QueryData('WaterLevel', 'branch1', 10)` is a valid query.
+    `QueryData('WaterLevel', 'branch1')` requests all the WaterLevel points
+    of `branch1`.
+    `QueryData('Discharge')` requests all the Discharge points of the model.
+    """
+
+    def __init__(self, variable_type, branch_name=None, chainage=None):
+        self._variable_type = variable_type
+        self._branch_name = branch_name
+        self._chainage = chainage
+        self._validate()
+
+    def _validate(self):
+        vt = self.variable_type
+        bn = self.branch_name
+        c = self.chainage
+        if not isinstance(vt, str):
+            raise TypeError("variable_type must be a string.")
+        if not vt in ["WaterLevel", "Discharge", "Pollutant"]:
+            raise ValueError(
+                f"Bad variable_type {vt} entered. "
+                "It must be either 'WaterLevel', 'Discharge' or 'Pollutant'."
+            )
+        if bn is not None and not isinstance(bn, str):
+            raise TypeError("branch_name must be either None or a string.")
+        if c is not None and not isinstance(c, (int, float)):
+            raise TypeError("chainage must be either None or a number.")
+        if bn is None and c is not None:
+            raise ValueError("chainage cannot be set if branch_name is None.")
+
+    @property
+    def variable_type(self):
+        return self._variable_type
+
+    @property
+    def branch_name(self):
```

Comment:
Nice. Properties is a better way.

Reply:
Yep I've chosen to use properties here so that the attributes have to be validated only once at the initialization.